test_case/src/send_payment.rs

```diff
@@ -214,72 +190,87 @@ pub fn distrubite_coins_and_cocurrency(
     let paid_amount = value_t!(send.value_of("pay"), f64).ok();
 
     if cycle_index.is_some() && paid_amount.is_none() {
-        for i in 0..cycle_index.unwrap() {
-            let tmp_ws = Arc::clone(&ws);
-            let tmp_test_wallets = Arc::clone(&test_wallets);
-
-            go!(move || {
-                let mut index = i;
-
-                REGISTERED_WALLETS.write().unwrap().insert(i);
-                loop {
-                    let tmp_ws = Arc::clone(&tmp_ws);
-
-                    if let Err(e1) = continue_sending(tmp_ws, index, &tmp_test_wallets) {
-                        eprintln!("wallet {} send payment error {} ", index, e1);
-                        index = match choose_wallet(index, &tmp_test_wallets) {
-                            Ok(v) => {
-                                println!("rechoose wallet {} ", v);
-                                v
-                            }
-                            Err(e2) => {
-                                coroutine::sleep(Duration::from_secs(10));
-                                eprintln!("fail to choose wallet {} ", e2);
-                                index
-                            }
-                        };
-                    };
-
-                    coroutine::yield_now();
-                    coroutine::sleep(Duration::from_millis(100));
-                }
-            });
-        }
+        share_token_in_wallets(&ws, cycle_index.unwrap(), test_wallets);
+    } else if let Some(num) = paid_amount {
+        distribute_token(
+            &ws,
+            &wallet_info,
+            num,
+            cycle_index.unwrap_or(1),
+            witnesses,
+            &test_wallets,
+        );
+    }
 
-        loop {
-            coroutine::sleep(Duration::from_secs(10));
-        }
+    Ok(())
+}
+
+pub fn share_token_in_wallets(
+    ws: &Arc<WalletConn>,
+    concurrent_counts: usize,
+    test_wallets: Vec<wallet::WalletInfo>,
+) {
+    let arc_test_wallets = Arc::new(test_wallets);
+    for i in 0..concurrent_counts {
+        let tmp_ws = Arc::clone(&ws);
+
+        let tmp = Arc::clone(&arc_test_wallets);
+        may::go!(move || {
+            let index = i;
+            let mut wallet_info = choose_wallet(i, &tmp).unwrap();
+
+            loop {

```

Comment:
move this API into the binary, this should not be in library.
libs should only contains kernel interfaces, and leave the business logic to user.